### PR TITLE
add horizontalpodautoscaler

### PR DESCRIPTION
**Pull request by coolbry95**
### Proposed changes
Adds the horizontalpodautoscaler for the ingress controller

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Review by shaun-nx — APPROVED** [review]

**Review by vepatel — APPROVED** [review]

File: deployments/helm-chart/values.yaml [hunk 1]
```diff
@@ -119,6 +119,22 @@ controller:
   ## The termination grace period of the Ingress Controller pod.
   terminationGracePeriodSeconds: 30
 
+
```
Comment (lucacome) [inline deployments/helm-chart/values.yaml:122]:
```suggestion
```

File: deployments/helm-chart/README.md [hunk 2]
```diff
@@ -248,6 +248,12 @@ Parameter | Description | Default
 `controller.readyStatus.initialDelaySeconds` | The number of seconds after the Ingress Controller pod has started before readiness probes are initiated. | 0
 `controller.enableLatencyMetrics` | Enable collection of latency metrics for upstreams. Requires `prometheus.create`. | false
 `controller.minReadySeconds` | Specifies the minimum number of seconds for which a newly created Pod should be ready without any of its containers crashing, for it to be considered available. [docs](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#min-ready-seconds) | 0
+`controller.autoscaling.enabled` | Enables HorizontalPodAutoscaling. | false
+`controller.autoscaling.annotations` | The annotations of the Ingress Controller HorizontalPodAutoscaler. | {}
+`controller.autoscaling.minReplicas` | Minimum number of replicas for the HPA. | 1
+`controller.autoscaling.maxReplicas` | Maximum number of replicas for the HPA. | 3
+`controller.autoscaling.targetCPUUtilizationPercentage`` | The target cpu utilization percentage. | 50
+`controller.autoscaling.targetMemoryUtilizationPercentage`` | The target memory utilization percentage. | 50
```
Comment (lucacome) [inline deployments/helm-chart/README.md:256]:
```suggestion
`controller.autoscaling.targetCPUUtilizationPercentage` | The target CPU utilization percentage. | 50
`controller.autoscaling.targetMemoryUtilizationPercentage` | The target memory utilization percentage. | 50
```

File: deployments/helm-chart/templates/controller-hpa.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    kind: Deployment
+    name: {{ default (include "nginx-ingress.name" .) .Values.controller.name }}
+  minReplicas: {{ .Values.controller.autoscaling.minReplicas }}
+  maxReplicas: {{ .Values.controller.autoscaling.maxReplicas }}
+  metrics:
+    {{- if .Values.controller.autoscaling.targetCPUUtilizationPercentage }}
+    - type: Resource
+      resource:
+        name: cpu
+        targetAverageUtilization: {{ .Values.controller.autoscaling.targetCPUUtilizationPercentage }}
```
Comment (lucacome) [inline deployments/helm-chart/templates/controller-hpa.yaml:25]:
```suggestion
```
This doesn't seem to be a field?

File: deployments/helm-chart/templates/controller-hpa.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        target:
+          type: Utilization
+          averageUtilization: {{ .Values.controller.autoscaling.targetCPUUtilizationPercentage }}
+    {{- end }}
+    {{- if .Values.controller.autoscaling.targetMemoryUtilizationPercentage }}
+    - type: Resource
+      resource:
+        name: memory
+        targetAverageUtilization: {{ .Values.controller.autoscaling.targetMemoryUtilizationPercentage }}
```
Comment (lucacome) [inline deployments/helm-chart/templates/controller-hpa.yaml:34]:
```suggestion
```
same as the one above, not a field

File: deployments/helm-chart/templates/controller-hpa.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{{- if and .Values.controller.autoscaling.enabled (or (eq .Values.controller.kind "deployment")) -}}
+apiVersion: autoscaling/v2
```
Comment (lucacome) [inline deployments/helm-chart/templates/controller-hpa.yaml:2]: `autoscaling/v2` is only available starting from k8s 1.23, so I think we need to check the version before deploying this

**Review by lucacome — CHANGES_REQUESTED** [review]

File: deployments/helm-chart/templates/controller-hpa.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+{{- if and .Values.controller.autoscaling.enabled (or (eq .Values.controller.kind "deployment")) -}}
```
Comment (lucacome) [inline deployments/helm-chart/templates/controller-hpa.yaml:1]:
```suggestion
{{- if and .Values.controller.autoscaling.enabled (eq .Values.controller.kind "deployment") -}}
```

File: deployments/helm-chart/templates/controller-hpa.yaml (hunk 5 above)
Comment (jasonwilliams14) [reply deployments/helm-chart/templates/controller-hpa.yaml:2]: @lucacome Yes I believe you are correct. We need to verify this.

Comment (coolbry95) [reply deployments/helm-chart/templates/controller-hpa.yaml:2]: That is why I left both types of resources. Wasn't sure how much templating we wanted to do to support older versions.

Comment (lucacome) [reply deployments/helm-chart/templates/controller-hpa.yaml:2]: what do you mean by both types?

Comment (jasonwilliams14) [reply deployments/helm-chart/templates/controller-hpa.yaml:2]: @coolbry95 can you clarify by both types?

Comment (coolbry95) [reply deployments/helm-chart/templates/controller-hpa.yaml:2]: Types is the wrong word. I should have said apiVersion. I think this is the simplest way though. Just support autoscaling/v2.

For example we could template and add conditionals to support autoscaling/v1 and autoscaling/v2 at the same time.

**Pull request closed without merging**